docs/proposals/20200511-clusterctl-extensible-template-processing.md

```diff
@@ -0,0 +1,344 @@
+  separately which disrupts the UX of clusterctl. If each infrastructure
+  provider provides different template styles, the user would need to install
+  multiple templating tool binaries.
+### Risks and Mitigations
+- Support path for supporting extended templating tools such as ytt within
+  clusterctl. That is, if there are templating issues, who owns them and how
+  do we address them. This is for each templating engine that is included in
+  clusterctl.  
+  However since we’ve decided to put extended templating
+  libraries within a **contrib** folder, it will be the responsibility of the
+  contributor and not the project maintainers to support the templating
+libraries so the issue of support should be solved with this contract.
+  
+- Each provider may want to implement their own templating standard. This
+  could affect the UX of clusterctl when automatically pulling workload
+  cluster templates from various providers.
+
+
+### Constraints
+- Currently, clusterctl relies on the conformance of file name conventions
+  such as `infrastructure-components.yaml` and
+  `cluster-template-<flavor>.yaml`. Other templating tools might require other
+  conventions to be defined and followed to allow the same “day 1” experience.
+- Some templating tools will require multiple files to be defined rather than
+  a single yaml file. These artifacts will need to be “grouped” together to
+  support current retrieval mechanisms. Currently, `clusterctl config cluster`
+  retrieves templates from multiple sources such as ConfigMaps within a
+  cluster, URL, Github Repository, Local Repository and even the overrides
+  directory. To ensure compatibility, we’ll need to establish a compression
+  format like tar, gzip or zip.
+
+## Alternatives
+- An alternative to allowing clusterctl handle templates would be to just use
+  another tool upstream to generate the various templates with variables
+  defined. Then use clusterctl to consume that template and use the default
+  variable substitution templating solution.
+
+
+## Upgrade Strategy
+
+
+Any individual template engine libraries would be upgraded via go modules
+preferably using semver standards.
+
+
+## Additional Details
+
+
+Currently, cluster-api-provider-azure is using kustomize to author their
+cluster template flavors. See [issue 540] for more details. Maybe we can
+collect input from them to understand their strategy if they need to
+manage their template variations.
+
+
+## Implementation History
+
+
+- [wfernandes] 04/22/2020: Proposed idea in an issue or [community meeting]
+- [wfernandes/fabrizio.pandini] 04/27/2020: Compile a [CAEP Google Doc] following the CAEP template
+- [wfernandes] 05/06/2020: First round of feedback from community
+- [wfernandes ] 05/13/2020: Present proposal at a [community meeting]
+- [wfernandes ] 05/11/2020: Open proposal PR
+
+
+## Open Questions
+<!-- Add your questions here -->
+
+1. [wfernandes] How do we pass in values to the YamlProcessor? Currently,
```

Comment:
This seems similar to the problem that say, [Puppet Hiera](https://puppet.com/docs/puppet/latest/hiera_intro.html), tried to solve with a pluggable model. Most frequent options I've seen are the desire to use GPG or a system like Hashicorp Vault for access to secrets. The current model of environment variables or values in clusterctl.yaml is a bit poor in that regards.

The AWS CLI recently added the ability to [shell out to an external program](https://puppet.com/docs/puppet/latest/hiera_intro.html) for the retrieval of values which is also a neat model.

Reply:
These are all good options that can be added in the future because one of the goals of this proposal is to not change how `clusterctl` reads in values.
After investigating for the specific use case of ytt, we can use the existing method of reading in values either from the `clusterctl` config file. Since `clusterctl` uses viper under the hood to read in the config file, it is able to read nested values as well. So for the example mentioned below, we can specify 
```
aws:
  secret_key: foo
  access_key: bar
``` 
in the `clusterctl` config file and clusterctl will read it in as `aws.secret_key`. So I believe this question can be resolved. 🙂